contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java

```diff
@@ -118,20 +137,58 @@ public Span handleStart(C carrier, Q request) {
     if (publicEndpoint && spanContext != null) {
       span.addLink(Link.fromSpanContext(spanContext, Type.PARENT_LINKED_SPAN));
     }
-    return span;
+
+    if (span.getOptions().contains(Options.RECORD_EVENTS)) {
+      addSpanRequestAttributes(span, request, extractor);
+    }
+
+    return getNewContext(span, tagger.getCurrentTagContext());
   }
 
   /**
-   * Close an HTTP span.
+   * Close an HTTP span and records stats specific to the request.
    *
-   * <p>This method will set status of the span and end it.
+   * <p>This method will set status of the span and end it. Additionally it will record message
+   * events for the span and record measurements associated with the request.
    *
+   * @param context the {@link HttpRequestContext} used with {@link
+   *     HttpServerHandler#handleStart(Object, Object)}
+   * @param request the HTTP request entity.
    * @param response the HTTP response entity. {@code null} means invalid response.
    * @param error the error occurs when processing the response.
-   * @param span the span.
    * @since 0.19
    */
-  public void handleEnd(Span span, @Nullable P response, @Nullable Throwable error) {
-    spanEnd(span, response, error);
+  public void handleEnd(
+      HttpRequestContext context, Q request, @Nullable P response, @Nullable Throwable error) {
+    recordStats(context, request, response, error);
+    spanEnd(context.span, response, error);
+  }
+
+  private void recordStats(
+      HttpRequestContext context, Q request, @Nullable P response, @Nullable Throwable error) {
+    checkNotNull(context, "context");
```

Comment:
This is not a public API, why do we do the checks here? I would suggest to move all the checks in the public API and all private APIs like recordStats and spanEnd should accept non-null values. This way we avoid duplicate checks and also we make the code cleaner.

Reply:
done.